plutus-pab/src/Plutus/PAB/Core/ContractInstance/STM.hs

```diff
@@ -394,31 +405,44 @@ insertInstance :: ContractInstanceId -> InstanceState -> InstancesState -> STM (
 insertInstance instanceID state (InstancesState m) = STM.modifyTVar m (Map.insert instanceID state)
 
 -- | Wait for the status of a transaction to change.
-waitForTxStatusChange :: TxStatus -> TxId -> BlockchainEnv -> STM TxStatus
+waitForTxStatusChange
+  :: TxStatus -> TxId -> BlockchainEnv -> STM TxStatus
 waitForTxStatusChange oldStatus tx BlockchainEnv{beTxChanges, beLastSyncedBlockNo} = do
-    txIdState   <- _usTxUtxoData . utxoState <$> STM.readTVar beTxChanges
-    blockNumber <- STM.readTVar beLastSyncedBlockNo
-    let txStatus = transactionStatus blockNumber txIdState tx
-    -- Succeed only if we _found_ a status and it was different; if
-    -- the status hasn't changed, _or_ there was an error computing
-    -- the status, keep retrying.
-    case txStatus of
-      Right s | s /= oldStatus -> pure s
-      _                        -> empty
+    case beTxChanges of
+      Left ix -> do
+        blockNumber <- STM.readTVar beLastSyncedBlockNo
+        txIdState <- _usTxUtxoData . utxoState <$> STM.readTVar ix
+        let txStatus  = transactionStatus blockNumber txIdState tx
+        -- Succeed only if we _found_ a status and it was different; if
+        -- the status hasn't changed, _or_ there was an error computing
+        -- the status, keep retrying.
+        case txStatus of
+          Right s | s /= oldStatus -> pure s
+          _                        -> empty
+      -- This branch gets intercepted in `processTxStatusChangeRequestIO` and
+      -- handled separateley, so we should never reach this place.
+      Right _ ->
```

Comment:
This is probably not my finest moment, but fixing it would mean refactoring quite a bit more code, and I am not sure that is warranted as the in-memory index should eventually go away which will make life a lot simpler.